Update to pyro 1.8.1

icfly2 opened the pull request:
I've started the process of updating the dependencies and added an example for classification.

Things changed to enable move to pyro-ppl 1.8.1:
- [x] fixed reference to `deep_setattr` in pyro.infer.autoguide.guides.deep_setattr
- [x] updated reference for `torch.cholesky` to `torch.linalg.cholesky`
- [x] added trange dependency for `fit` in svi to bring it in line with mcmc behaviour from pyro `tqdm` is a pyro dependency, so no new dependency is added to TyXe. 
- [x]  figured out why `vs` in `tests/test_bnn.py:69` has changed value (to look more like old `vs` + `torch.eye`)
- [x] fixed error in example/classification.py for `likelihood.data shape`
- [x] check that all examples still work (some don't work for me with pyro==1.4.0)

bnn.py:207 triggers 
```
ValueError: at site "likelihood.data", invalid log_prob shape
  Expected [20], actual [20, 20]
  Try one of the following fixes:
  - enclose the batched tensor in a with pyro.plate(...): context
  - .to_event(...) the distribution being sampled
  - .permute() data dimensions
```
I haven't found what it refers to as when i step through it, all `pyro.sample` statements are in plates. The relevant example in pyro `eight_schools` has also been updated for the newer versions of pyro to include these `pyro.plate` contexts.
 
I'd appreciate input on the approach for the classification model as well as help with fixing these last two errors.

icfly2 commented:
I have made a few decisions regarding the new names of the pyro parameters. It adds `net.` to the parameters, but not for all methods invoke.

The `vcl.py` example fails when the prior is updated to the `DictPrior`. Either a workaround can try names or net. + names or the example needs more adjustments.

There is also a script to test all the examples (bar two) (but not all options) to make sure they run for future development. Please have a look and let me know where you disagree with the direction. And any help on the bnn test would be much appreciated.

icfly2 commented:
So a bit more digging and it turns out that net_guide.scale_tril returns a unitriangular matrix rather than a lower triangular in the newer versions of pyro. This is not due to the change from ` torch.cholesky` to `torch.linalg.cholesky` as I've checked and they return the same as expected and documented.

This can be remedied by multiplying the result with the scale vector, recreating the old triangular matrix. I don't know where else this change would have implications and I've not found the part in the pyro changelog where this change was documented.

Tests now pass. Last remaining issue is the vsl example's dictprior naming and I'm tempted to leave that for now and fix it after the PR.

icfly2 commented on examples/resnet.py:
```diff
@@ -98,13 +98,13 @@ def main(dataset, architecture, inference, train_batch_size, test_batch_size, lo
         if inference == "last-layer-mean-field":
             guide = functools.partial(tyxe.guides.AutoNormal,
-                                      init_loc_fn=tyxe.guides.PretrainedInitializer.from_net(net), init_scale=1e-4)
+                                      init_loc_fn=tyxe.guides.PretrainedInitializer.from_net(net, prefix="net"), init_scale=1e-4)
```
I have not been able to test this, as I'm not sure how the pretrained weights are to be passed to the model.

icfly2 commented on examples/test_examples.sh:
```diff
@@ -0,0 +1,21 @@
+python resnet.py --inference map --num-epochs 10
+
+# Adding a test for one of the last-laye rones would be nice, but not sure how to handle the shipping of pre-trained weights
+# python resnet.py --inference last-layer-mean-field --num-epochs 10
```
See comment in resnet.py file

ucl-jhr replied:
We could take the initial random ones in the script since it's just for testing that the script doesn't crash.

icfly2 commented on tests/test_bnn.py:
```diff
@@ -56,22 +56,39 @@ def test_diagonal_svi():
 
     vm = pyro.get_param_store()["net_guide.loc"].data.squeeze()
-    vs = pyro.get_param_store()["net_guide.scale_tril"].data
+    vsd = pyro.get_param_store()["net_guide.scale"].data # This value has changed!!
```
Code comments to be removed after review.

icfly2 commented on tyxe/priors.py:
```diff
@@ -189,8 +192,12 @@ def __init__(self, prior_dict, *args, **kwargs):
     def prior_dist(self, name, module, param):
-        return self.prior_dict[name]
-
+        # I find this preferable to the hack in update_ as it gets to the root. 
```
Unless the the naming of the keys can be fixed in pyro, this strikes me as the nicest solution. Rather than the hack in line 142

icfly2 commented on tyxe/priors.py:
```diff
@@ -139,6 +139,9 @@ def update_(self, net):
         for module_name, module in net.named_modules():
             for site_name, site in list(util.named_pyro_samples(module, recurse=False)):
                 full_name = module_name + "." + site_name
+                # See change in DictPrior as an alternative
```
Solution below is preferable in my opinion. Comments to be removed after review.

icfly2 commented on tyxe/util.py:
```diff
@@ -9,18 +9,11 @@
-deep_setattr = pyro.infer.autoguide.guides._deep_setattr
-deep_getattr = pyro.util.deep_getattr
-
+try:
```
If only one version of pyro is to be supported, then this can be simplified.

ucl-jhr replied:
Yes I'd go with only supporting the most recent pyro version. We could just wrap their `deep_attr` function with a deprecation warning (see [here](https://github.com/pyro-ppl/pyro/blob/d7687ae0f738bd81a792dabbb18a53c0fce73765/pyro/generic.py#L9) for an example from the pyro codebase) and eventually remove them here alltogether.

icfly2 commented:
I've done my best to address your comments.

See the comments in the files for the remaining issues.

I'll contribute the github actions for release to pypi and testing after this PR, if you like.

ucl-jhr commented on examples/vcl.py:
```diff
@@ -173,14 +185,15 @@ def callback(_i, _ii, e):
     parser.add_argument("--root", default=ROOT)
     parser.add_argument("--dataset", choices=["mnist", "cifar"], required=True)
     parser.add_argument("--inference", choices=["mean-field", "ml"], required=True)
+    parser.add_argument("--test", action='store_true')
```
I think I'd rather add an explicit `num-epochs` argument that is by default `60` for CIFAR and `600` for MNIST since that is generally useful and more explicit as to what it does based on the name.

ucl-jhr commented on tests/test_bnn.py:
```diff
@@ -1,4 +1,5 @@
 from functools import partial
+from packaging import version
```
Is this still needed somewhere?

ucl-jhr replied:
Ah sorry, misread the end of the file. But as per my earlier comment, I'd remove this and assume pyro version 1.8.1.

ucl-jhr commented on tyxe/bnn.py:
```diff
@@ -200,7 +200,7 @@ def fit(self, data_loader, optim, num_epochs, callback=None, num_particles=1, cl
         svi = SVI(self.model, self.guide, optim, loss=loss)
 
-        for i in range(num_epochs):
+        for i in trange(num_epochs):
```
I wouldn't use `tqdm` by default, if someone wants a progress bar they can handle that through the callback.

icfly2 replied:
I added it as it is a pyro dependency, so should be on the users machine. But, yes can easily be removed. Maybe, a few common nice callbacks can be added to the library for convenience.

ucl-jhr commented on tyxe/guides.py:
```diff
@@ -114,22 +120,22 @@ def _setup_prototype(self, *args, **kwargs):
             if self.train_loc:
                 unconstrained_value = pynn.PyroParam(unconstrained_value)
-            ag.guides._deep_setattr(self, name + ".loc", unconstrained_value)
+            util.deep_setattr(self, name + ".loc", unconstrained_value)
```
I'd handle `deep_setattr` as `deep_getattr` and use the `pyro` function, then we don't need to keep them around in the tyxe codebase.

ucl-jhr commented:
Left you a few comments. Thanks again for the updates. The big TODOs left at this point for me are:

- [ ] Remove the new examples to submit them in a separate PR. I'd prefer to keep the changes minimal and focused in a single PR to make it easier to rollback in case anything does break.
- [ ] Remove the examples execution script for the same reason for now. Happy to set up continuous integration after this PR is merged.
- [ ] Remove bits of code that account for the possibility of different pyro versions and assume 1.8.1.
- [ ] Add deprecation warnings for `deep_getattr` and `deep_settattr` in `util` and use the pyro functions in the rest of the code base.
- [ ] Solve the prefix issue for the continual learning case. This one might require a little bit more discussion.

Let me know if I forgot anything :)

Btw should I have access to your branch? I think there's a github option to set up PRs in that way. In that case I can obviously make some code changes as well.

icfly2 commented:
> Left you a few comments. Thanks again for the updates. The big TODOs left at this point for me are:

Thanks a lot, i'll see what I can chew through in the next few days, most are fairy minor. I agree with all your points.
 
> Btw should I have access to your branch? I think there's a github option to set up PRs in that way. In that case I can obviously make some code changes as well.

the branch isn't protected, but I guess the way GitHub works you'd have to make PRs against my branch. If this turns into a continuing effort, I guess I could be added as a collaborator on the original repo and then make branches there. But if you know of a setting I'm missing let me know, I'm not that familiar with GitHub.

icfly2 commented:
Ok I think I've adressed everything i can. Only the deprecation warnings are left adn then the actualy push to PyPi but i guess they  are for after the PR is merged. Let me know if there is anything else I've missed.

ucl-jhr commented:
Thanks could you just also remove the regression example?

Also for the prefix, I meant having it as an argument on the `get_detached_distributions` of the `AutoNormal` autoguide class. So if you could move that functionality there and revert the changes to `DictPrior` that would be great. Sorry if I hadn't been clear/am being pedantic, but I think it makes a lot more sense to pull these distribution names correctly out of the guide in case anyone wants to use them downstream. And have a look at [how pytorch handles prefixes](https://github.com/pytorch/pytorch/blob/893d7632761e9cd34ac92dea316090f3847d196c/torch/nn/modules/module.py#L1660), there's no need to include a `"."` in the string just add it when concatenating it with the name -- makes calling the function easier by one character :)

Other little bits to clean up (I can handle those as well after merging):
- [ ] remove the `--test` argument in the `vcl.py` example and replace it with a `--num-epochs` one.
- [ ] remove the comments in the `multivariate_svi` test.
- [ ] revert the change from `range` to `trange` in the bnn `fit` loop.
- [ ] change occurrences of `util.deep_setattr` to `pyro.infer.autoguide.utils.deep_setattr`.
- [ ] revert removal of `deep_getattr` and `deep_hasattr` (breaks backward compatibiltiy).
- [ ] add deprecation warning to all `deep_attr` functions.

icfly2 commented:
- [x] remove the regression example.
>     * [x]  remove the `--test` argument in the `vcl.py` example and replace it with a `--num-epochs` one.
> 
>     * [x]  remove the comments in the `multivariate_svi` test.
> 
>     * [x]  revert the change from `range` to `trange` in the bnn `fit` loop.
> 
>     * [ ]  change occurrences of `util.deep_setattr` to `pyro.infer.autoguide.utils.deep_setattr`.
> 
>     * [ ]  revert removal of `deep_getattr` and `deep_hasattr` (breaks backward compatibiltiy).
> 
>     * [ ]  add deprecation warning to all `deep_attr` functions.

Not 100% sure I follow. These functions are only accessed internally in tyxe functions and the import in utils is just a renaming. Unless you want to add the check for the version of pyro back, this won't run anyway. Also as there is no PyPi package, I'm not sure how you would push the deprecation warning to users before it is actually deprecated. but maybe I'm jst misunderstanding the whole thing, in which case please just change as you see fit.

icfly2 commented:
Other than the depreciation warning, where I don't quite understand what you're trying to achieve, I think I have addressed all your points. I appreciate the detailed, nitpicking feedback, that is how things progress well.

ucl-jhr commented:
Cool, looks good to merge now. I was playing around with the whole prefix issue for the prior update and found a nicer way of fixing this directly in the BNN class, I'll push that after merging your PR (I was actually wrong that the issue came from the guide, the net prefix is handled correctly there out of the box now).

Regarding the deprecation warning: I don't want to remove public facing functions without a warning. I don't know if anyone might be using them and if they would pull updates to the codebase, it just seems like the safer choice to me than outright deleting them.

Glad you appreciate the feedback, I hope it wasn't too much back and forth about small details!
